formatprocessor/formatprocessor.go

```diff
@@ -0,0 +1,532 @@
+
+	// returns the PTS of the unit.
+	GetPTS() time.Duration
+}
+
+// H264 is a H264 data unit.
+type H264 struct {
+	Base
+	AU [][]byte
+}
+
+// Base contains fields shared across all units.
+type Base struct {
+	RTPPackets []*rtp.Packet
+	NTP        time.Time
+	PTS        time.Duration
+}
+
+// GetRTPPackets implements Unit.
+func (u *Base) GetRTPPackets() []*rtp.Packet {
+	return u.RTPPackets
+}
+
+// GetNTP implements Unit.
+func (u *Base) GetNTP() time.Time {
+	return u.NTP
+}
+
+// GetPTS implements Unit.
+func (u *Base) GetPTS() time.Duration {
+	return u.PTS
+}
+
+// Processor processes RTP packets & turns them into Units.
+type Processor interface {
+	// process a Unit.
+	ProcessUnit(Unit) error
+
+	// process a RTP packet and convert it into a unit.
+	ProcessRTPPacket(
+		pkt *rtp.Packet,
+		ntp time.Time,
+		pts time.Duration,
+		hasNonRTSPReaders bool,
+	) (Unit, error)
+}
+
+// New returns a new Processor.
+func New(
+	udpMaxPayloadSize int,
+	forma format.Format,
+	generateRTPPackets bool,
+) (Processor, error) {
+	switch forma := forma.(type) {
+	case *format.H264:
+		return newH264(udpMaxPayloadSize, forma, generateRTPPackets)
+
+	default:
+		return nil, errors.New("unsupported formatprocessor")
+	}
+}
+
+// extract SPS and PPS without decoding RTP packets.
+func rtpH264ExtractParams(payload []byte) ([]byte, []byte) {
+	if len(payload) < 1 {
+		return nil, nil
+	}
+
+	typ := h264.NALUType(payload[0] & 0x1F)
+
+	switch typ {
+	case h264.NALUTypeSPS:
+		return payload, nil
+
+	case h264.NALUTypePPS:
+		return nil, payload
+
+	case h264.NALUTypeSTAPA:
+		payload := payload[1:]
+		var sps []byte
+		var pps []byte
+
+		for len(payload) > 0 {
+			if len(payload) < 2 {
+				break
+			}
+
+			size := uint16(payload[0])<<8 | uint16(payload[1])
+			payload = payload[2:]
+
+			if size == 0 {
+				break
+			}
+
+			if int(size) > len(payload) {
+				return nil, nil
+			}
+
+			nalu := payload[:size]
+			payload = payload[size:]
+
+			typ = h264.NALUType(nalu[0] & 0x1F)
+
+			switch typ {
+			case h264.NALUTypeSPS:
+				sps = nalu
+
+			case h264.NALUTypePPS:
+				pps = nalu
+
+			case h264.NALUTypeNonIDR,
+				h264.NALUTypeDataPartitionA,
+				h264.NALUTypeDataPartitionB,
+				h264.NALUTypeDataPartitionC,
+				h264.NALUTypeIDR,
+				h264.NALUTypeSEI,
+				h264.NALUTypeAccessUnitDelimiter,
+				h264.NALUTypeEndOfSequence,
+				h264.NALUTypeEndOfStream,
+				h264.NALUTypeFillerData,
+				h264.NALUTypeSPSExtension,
+				h264.NALUTypePrefix,
+				h264.NALUTypeSubsetSPS,
+				h264.NALUTypeReserved16,
+				h264.NALUTypeReserved17,
+				h264.NALUTypeReserved18,
+				h264.NALUTypeSliceLayerWithoutPartitioning,
+				h264.NALUTypeSliceExtension,
+				h264.NALUTypeSliceExtensionDepth,
+				h264.NALUTypeReserved22,
+				h264.NALUTypeReserved23,
+				h264.NALUTypeSTAPB,
+				h264.NALUTypeMTAP16,
+				h264.NALUTypeMTAP24,
+				h264.NALUTypeFUA,
+				h264.NALUTypeSTAPA,
+				h264.NALUTypeFUB:
+			}
+		}
+
+		return sps, pps
+	case h264.NALUTypeNonIDR,
```

Comment:
Talked offline about how the exhaustive case statements here are indeed useful. If the gortsplib api changes, the linter could flag this pointing us to important changes.

Reply:
I added this branch to the case statement so that when we have exhaustive case statement linting enabled (as we do in RRDK) if if we update the `h264` and it adds a new type, we will need to handle that case & won't silently do the wrong thing.

I agree it adds verbosity but think it is worth it to get feedback if our dependencies change out from underneath us.